### ISSUE TITLE
When using OVN on IPv4/IPv6 dual stack the MAC address of the IPv4 gateway is not learnt.

### ISSUE THREAD

**Issue by iherdgoats** [lxc/lxd#11492]
# Issue description

When using LXD and OVN on a dual stack uplink network the MAC address of the IPv4 gateway is not learnt. This leads to broken northbound connectivity. However we do have a mac binding for the IPv6 gateway,

Looking in the OVN driver code I see there's a workaround that makes the driver issue an ICMPv6 request in order to discover the IPv6 gateway but there is no corresponding IPv4 workaround.

# Steps to reproduce

 1. Deploy LXD on a dual stack IPv4/IPv6 network.
 2. Install and run the OVN chassis on all LXD nodes.
 3. Create an uplink network attached to a physical NIC on the IPv4/IPv6 network.
 4. Create an OVN network that uses the uplink as a parent.
 5. Launch a container and attempt to ping a public IP or the next hop.
 6. No response will be received.
 7. Check the SB MAC Binding database and there will be no entry.
 8. Ping the routers external IPv4 address.
 9. A MAC binding will be created and IPv4 internet connectivity will be established.

# Information to attach

LXD version:  5.12-c63881f
OVN version: 22.03.0

```
$ microovn.ovn-nbctl show 
switch 0d8be25c-8503-4989-8af0-76f4d408313b (lxd-net2-ls-int)
    port lxd-net2-instance-c0890aa9-a87a-4a4a-b184-6f5d6ec3b04e-eth0
        addresses: ["00:16:3e:b0:55:c4 dynamic"]
    port lxd-net2-ls-int-lsp-router
        type: router
        router-port: lxd-net2-lr-lrp-int
switch 47f2de8f-e5f9-47b4-ba6b-4a4dba63f9ee (lxd-net2-ls-ext)
    port lxd-net2-ls-ext-lsp-router
        type: router
        router-port: lxd-net2-lr-lrp-ext
    port lxd-net2-ls-ext-lsp-provider
        type: localnet
        addresses: ["unknown"]
router 3354b1b6-1ae4-4de5-b2d7-e5c189474137 (lxd-net2-lr)
    port lxd-net2-lr-lrp-int
        mac: "00:16:3e:f8:47:c0"
        networks: ["10.7.224.1/24", "fd42:3db6:cbae:c768::1/64"]
    port lxd-net2-lr-lrp-ext
        mac: "00:16:3e:f8:47:c0"
        networks: ["10.1.0.100/24", "fd55:13f7:9dcd:1::100/64"]
    nat 1b5b525d-1e4f-43ef-81e2-f7040b3a75a3
        external ip: "10.1.0.100"
        logical ip: "10.7.224.0/24"
        type: "snat"
    nat 6ad0727b-c01c-415a-9208-096abc31ee08
        external ip: "fd55:13f7:9dcd:1::100"
        logical ip: "fd42:3db6:cbae:c768::/64"
        type: "snat"
```

```
$ microovn.ovn-sbctl list MAC_Binding
_uuid               : abcf63c5-dee7-4295-ae63-ad7381fa12ef
datapath            : d3217d6b-1e26-43e1-8e08-fec0c44e9e78
ip                  : "::"
logical_port        : lxd-net2-lr-lrp-int
mac                 : "00:00:00:00:00:00"

_uuid               : cb41e19e-b5f0-4110-a82c-73703ae80917
datapath            : d3217d6b-1e26-43e1-8e08-fec0c44e9e78
ip                  : "fe80::f816:3eff:fead:e264"
logical_port        : lxd-net2-lr-lrp-ext
mac                 : "fa:16:3e:ad:e2:64"

_uuid               : 4e5263a7-4491-48ad-8c14-b82b133ae5ab
datapath            : d3217d6b-1e26-43e1-8e08-fec0c44e9e78
ip                  : "fd55:13f7:9dcd:1::1"
logical_port        : lxd-net2-lr-lrp-ext
mac                 : "fa:16:3e:ad:e2:64"
```

**Comment by tomponline** [comment]
> Looking in the OVN driver code I see there's a workaround that makes the driver issue an ICMPv6 request in order to discover the IPv6 gateway but there is no corresponding IPv4 workaround.

We've never needed that before. So unless its a new regression in OVN I suspect its a misconfiguration or feature limitation of OVN.

**Comment by iherdgoats** [comment]
```
$ lxc network show uplink
config:
  dns.nameservers: 1.1.1.1,2606:4700::1111
  ipv4.gateway: 10.1.0.1/24
  ipv4.ovn.ranges: 10.1.0.100-10.1.0.253
  ipv4.routes: 10.1.0.0/24
  ipv6.gateway: fd55:13f7:9dcd:1::1/64
  ipv6.ovn.ranges: fd55:13f7:9dcd:1::100-fd55:13f7:9dcd:1::253
  ipv6.routes: fd55:13f7:9dcd:1::/64
  volatile.last_state.created: "false"
description: ""
name: uplink
type: physical
used_by:
- /1.0/networks/default
managed: true
status: Created
locations:
- lxd-virtual-edge-0
- lxd-virtual-edge-1
- lxd-virtual-edge-2
```

```
$ lxc network show default
config:
  bridge.mtu: "1422"
  ipv4.address: 10.7.224.1/24
  ipv4.nat: "true"
  ipv6.address: fd42:3db6:cbae:c768::1/64
  ipv6.nat: "true"
  network: uplink
  volatile.network.ipv4.address: 10.1.0.100
  volatile.network.ipv6.address: fd55:13f7:9dcd:1::100
description: ""
name: default
type: ovn
used_by:
- /1.0/instances/bold-goat
managed: true
status: Created
locations:
- lxd-virtual-edge-0
- lxd-virtual-edge-1
- lxd-virtual-edge-2
```

```
lxc info bold-goat
Name: bold-goat
Status: RUNNING
Type: container
Architecture: x86_64
Location: lxd-virtual-edge-0
PID: 4114
Created: 2023/03/20 19:31 UTC
Last Used: 2023/03/20 19:31 UTC

Resources:
  Processes: 57
  Disk usage:
    root: 1.40GiB
  CPU usage:
    CPU usage (in seconds): 138
  Memory usage:
    Memory (current): 1.01GiB
    Memory (peak): 1.25GiB
  Network usage:
    eth0:
      Type: broadcast
      State: UP
      Host interface: veth321964e4
      MAC address: 00:16:3e:b0:55:c4
      MTU: 1422
      Bytes received: 99.31MB
      Bytes sent: 674.90kB
      Packets received: 21391
      Packets sent: 8391
      IP addresses:
        inet:  10.7.224.2/24 (global)
        inet6: fd42:3db6:cbae:c768:216:3eff:feb0:55c4/64 (global)
        inet6: fe80::216:3eff:feb0:55c4/64 (link)
    lo:
      Type: loopback
      State: UP
      MTU: 65536
      Bytes received: 28.03kB
      Bytes sent: 28.03kB
      Packets received: 317
      Packets sent: 317
      IP addresses:
        inet:  127.0.0.1/8 (local)
        inet6: ::1/128 (local)
```

```
lxc exec bold-goat -- bash
root@bold-goat:~# ping 1.1.1.1
PING 1.1.1.1 (1.1.1.1) 56(84) bytes of data.
^C
--- 1.1.1.1 ping statistics ---
3 packets transmitted, 0 received, 100% packet loss, time 2048ms

root@bold-goat:~# 
```

**Comment by iherdgoats** [comment]
Basically OVN learns the IPv6 gateway (at fd55:13f7:9dcd:1::1/64) MAC address, but that may be due to router advertisements.

However OVN never learns the IPv4 gateway (at 10.1.0.1/24) MAC address. Unfortunately we have to use dual stack due a lack of northbound IPv6 connectivity from this environment.

**Comment by tomponline** [comment]
IPv4 discovery of the uplink gateway does work normally.

> ipv4.routes: 10.1.0.0/24

That looks suspect. You're indicating that the 10.1.0.0/24 subnet can be used by OVN networks downstream, but this isn't valid because its the same network as the uplink, and includes the uplink gateway.

**Comment by iherdgoats** [comment]
However running `ping 10.1.0.100` from the LXD host adds the following MAC binding:

```
$ microovn.ovn-sbctl list MAC_Binding
_uuid               : bdebead3-8468-4761-90dc-fc853a963e4f
datapath            : d3217d6b-1e26-43e1-8e08-fec0c44e9e78
ip                  : "10.1.0.1"
logical_port        : lxd-net2-lr-lrp-ext
mac                 : "fa:16:3e:63:a1:f1"
```

And northbound connectivity is established from the container.

**Comment by iherdgoats** [comment]
> Try taking those out and see if it helps.

Will try, thanks for the heads up

**Comment by iherdgoats** [comment]
Trying to remember why I set the routes in the first place, I believe this was for loadbalancers.

**Comment by tomponline** [comment]
Did it fix it?

You might need to use multiple smaller route subnets, rather than assigning the whole one.
If it fixes it, we should probably add an overlap check to prevent assigned downstream routes from covering the uplink gateway.

**Comment by iherdgoats** [comment]
> Did it fix it?

Just rebuilding the environment atm, so I can start from a clean state.

**Comment by iherdgoats** [comment]
Issue remains:

```
$ lxc network show uplink 
config:
  dns.nameservers: 1.1.1.1,2606:4700::1111
  ipv4.gateway: 10.1.0.1/24
  ipv4.ovn.ranges: 10.1.0.100-10.1.0.253
  ipv6.gateway: fd55:13f7:9dcd:1::1/64
  ipv6.ovn.ranges: fd55:13f7:9dcd:1::100-fd55:13f7:9dcd:1::253
  volatile.last_state.created: "false"
description: ""
name: uplink
type: physical
used_by:
- /1.0/networks/default
managed: true
status: Created
locations:
- lxd-virtual-edge-1
- lxd-virtual-edge-2
- lxd-virtual-edge-0
```

```
$ lxc network show default 
config:
  bridge.mtu: "1422"
  ipv4.address: 10.237.157.1/24
  ipv4.nat: "true"
  ipv6.address: fd42:5964:71c2:3d43::1/64
  ipv6.nat: "true"
  network: uplink
  volatile.network.ipv4.address: 10.1.0.100
  volatile.network.ipv6.address: fd55:13f7:9dcd:1::100
description: ""
name: default
type: ovn
used_by:
- /1.0/instances/helped-caribou
managed: true
status: Created
locations:
- lxd-virtual-edge-0
- lxd-virtual-edge-1
- lxd-virtual-edge-2
```

```
$ lxc exec helped-caribou -- bash
root@helped-caribou:~# ping 1.1.1.1
PING 1.1.1.1 (1.1.1.1) 56(84) bytes of data.
^C
--- 1.1.1.1 ping statistics ---
3 packets transmitted, 0 received, 100% packet loss, time 2031ms

root@helped-caribou:~# 
```

```
$ microovn.ovn-nbctl show
switch 3aed55f6-e3c8-4077-bf9a-86fcf4a0d5fa (lxd-net2-ls-int)
    port lxd-net2-ls-int-lsp-router
        type: router
        router-port: lxd-net2-lr-lrp-int
    port lxd-net2-instance-72cdf500-5f1a-44b7-9fbe-be601ee8e82e-eth0
        addresses: ["00:16:3e:a0:6a:f0 dynamic"]
switch 01705e83-2e8c-448b-a0f2-709e76aef3fb (lxd-net2-ls-ext)
    port lxd-net2-ls-ext-lsp-router
        type: router
        router-port: lxd-net2-lr-lrp-ext
    port lxd-net2-ls-ext-lsp-provider
        type: localnet
        addresses: ["unknown"]
router 99806d9c-82eb-4624-8efd-81adc0d07d34 (lxd-net2-lr)
    port lxd-net2-lr-lrp-int
        mac: "00:16:3e:24:25:12"
        networks: ["10.237.157.1/24", "fd42:5964:71c2:3d43::1/64"]
    port lxd-net2-lr-lrp-ext
        mac: "00:16:3e:24:25:12"
        networks: ["10.1.0.100/24", "fd55:13f7:9dcd:1::100/64"]
    nat 7a7fc07c-5822-4a42-b557-250bc1002c03
        external ip: "10.1.0.100"
        logical ip: "10.237.157.0/24"
        type: "snat"
    nat 7eaf9134-ed77-42e5-8952-f46b55a437de
        external ip: "fd55:13f7:9dcd:1::100"
        logical ip: "fd42:5964:71c2:3d43::/64"
        type: "snat"
```

```
$ microovn.ovn-sbctl list MAC_Binding
_uuid               : 879563ff-de61-4857-ab67-7906ed152db8
datapath            : dd678903-afe8-469b-b2e4-6fe422b8c8dd
ip                  : "::"
logical_port        : lxd-net2-lr-lrp-int
mac                 : "00:00:00:00:00:00"
```

**Comment by tomponline** [comment]
Please can you use tcpdump on the uplink port to check if ovn is sending ARP requests?

**Comment by iherdgoats** [comment]
Noticed something weird, if I ping the IPv6 gateway address with the empty MAC_Binding table,

```
$ ping6 fd55:13f7:9dcd:1::1
PING fd55:13f7:9dcd:1::1(fd55:13f7:9dcd:1::1) 56 data bytes
64 bytes from fd55:13f7:9dcd:1::1: icmp_seq=1 ttl=63 time=19.9 ms
^C
--- fd55:13f7:9dcd:1::1 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 19.935/19.935/19.935/0.000 ms
```

OVN sends out a neighbour solicitation, and we learn the gateway address.

```
$ tcpdump -i ens4
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on ens4, link-type EN10MB (Ethernet), capture size 262144 bytes
09:00:57.036939 IP6 fe80::f816:3eff:fe0a:dae3 > ip6-allnodes: ICMP6, router advertisement, length 56
09:01:03.155484 IP6 fe80::216:3eff:fe24:2512 > ff02::1:ff00:1: ICMP6, neighbor solicitation, who has fd55:13f7:9dcd:1::1, length 32
09:01:03.166733 IP6 fd55:13f7:9dcd:1::1 > ff02::1:ff00:100: ICMP6, neighbor solicitation, who has fd55:13f7:9dcd:1::100, length 32
^C
3 packets captured
3 packets received by filter
0 packets dropped by kernel
```

But pinging the IPv4 gateway results in failure.

```
$ ping 10.1.0.1
PING 10.1.0.1 (10.1.0.1) 56(84) bytes of data.
--- 10.1.0.1 ping statistics ---
4 packets transmitted, 0 received, 100% packet loss, time 3047ms
```

And no ARP:

```
sudo tcpdump -i ens4
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on ens4, link-type EN10MB (Ethernet), capture size 262144 bytes
^C
0 packets captured
0 packets received by filter
0 packets dropped by kernel
```

**Comment by tomponline** [comment]
I experience something similar today on a fresh setup, seems like it may be an OVN bug.

**Comment by tomponline** [comment]
@iherdgoats are you still experiencing this issue?

**Comment by iherdgoats** [comment]
I've since moved onto other projects, so cannot confirm the fix.

**Comment by tomponline** [comment]
Thanks. I've got something similar that @stgraber is experiencing with microcloud, although again, I am unable to reproduce, so will leave this open until we figure it out.

**Comment by tomponline** [comment]
@fnordahl identified an issue in OVS that should be fixed in the next OVS 2.17 point release:

https://github.com/openvswitch/ovs/commit/182b9cb3524c5e7be20be7ebf061c6730a7f7e26
